src/test/java/org/apache/commons/collections4/iterators/CartesianProductIteratorTest.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.collections4.iterators;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * Test class for CartesianProductIterator.
+ */
+public class CartesianProductIteratorTest extends AbstractIteratorTest<List<Character>> {
+
+    protected List<Character> letters;
+    protected List<Character> numbers;
+    protected List<Character> empty;
+
+    public CartesianProductIteratorTest() {
+        super(CartesianProductIteratorTest.class.getSimpleName());
+    }
+
+    @Override
+    public CartesianProductIterator<Character> makeEmptyIterator() {
+        return new CartesianProductIterator<>();
+    }
+
+    @Override
+    public CartesianProductIterator<Character> makeObject() {
+        return new CartesianProductIterator<>(letters, numbers);
+    }
+
+    @BeforeEach
+    public void setUp() {
+        letters = new ArrayList<>();
+        letters.add('A');
+        letters.add('B');
+        letters.add('C');
+        numbers = new ArrayList<>();
+        numbers.add('1');
+        numbers.add('2');
+        numbers.add('3');
+        empty = new ArrayList<>();
+    }
+
+    @Override
+    public boolean supportsRemove() {
+        return false;
+    }
+
+    @Test
+    public void testEmptyCollection() {
+        final CartesianProductIterator<Character> it = new CartesianProductIterator<>(letters, empty);
+        assertFalse(it.hasNext());
+    }
+
+    /**
+     * test checking that all the tuples are returned
+     */
+    @Test
+    public void testCartesianProductExhaustivity() {
```

Comment:
The test for the total output length should be supplemented with a test that the output is the desired Cartesian product:
```
[A, 1]
[A, 2]
[A, 3]
[B, 1]
[B, 2]
[B, 3]
[C, 1]
[C, 2]
[C, 3]
```

Reply:
Since this is meant to be like a nested for loop, let's test it like one:
```java
        final CartesianProductIterator<Character> it = makeObject();
        for (final Character a : letters) {
            for (final Character b : numbers) {
                Assertions.assertEquals(Arrays.asList(a, b), it.next());
            }
        }
        assertThrows(NoSuchElementException.class, () -> it.next());
```
You could even add another test with a third iterable to have a triple loop to demonstrate it works for increasing depth.